INT-3654: Rework TCP Connection Close

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3654

Previously, single-use `TcpConnections` self-closed when their use was complete.
This is unnatural and caused issues such as INT-3722.

Remove the self-closing behavior; connection close is now (properly) the responsibility
of the client using the connection:
- Client Side:
  -- `TcpOutboundGateway` after the reply is received
  -- `TcpSendingMessageHandler` after the send, when there is no collaborating inbound adapter
  -- `TcpReceivingChannelAdapter` when it is collaborating (after receiving the reply)
- Server Side:
  -- `TcpInboundGateway` after the reply is sent
  -- `TcpReceivingChannelAdapter` after the receive, when there is no collaborating outbound adapter
  -- `TcpSendingMessageHandler` when it is colllaborating (after sending the reply)

As before, the `CachingClientConnectionFactory` always sets single use on the target factory
to force it to create new connections on demand. It is always a single-use factory itself
so the clients return the connections to the pool (via `close()`).

**Needs a migration guide entry**

garyrussell committed:
INT-3654: Rework TCP Connection Close

JIRA: https://jira.spring.io/browse/INT-3654

Previously, single-use `TcpConnections` self-closed when their use was complete.
This is unnatural and caused issues such as INT-3722.

Remove the self-closing behavior; connection close is now (properly) the responsibility
of the client using the connection:

- Client Side:
-- `TcpOutboundGateway` after the reply is received
-- `TcpSendingMessageHandler` after the send, when there is no collaborating inbound adapter
-- `TcpReceivingChannelAdapter` when it is collaborating (after receiving the reply)

- Server Side:
-- `TcpInboundGateway` after the reply is sent
-- `TcpReceivingChannelAdapter` after the receive, when there is no collaborating outbound adapter
-- `TcpSendingMessageHandler` when it is colllaborating (after sending the reply)

As before, the `CachingClientConnectionFactory` always sets single use on the target factory
to force it to create new connections on demand. It is always a single-use factory itself
so the clients return the connections to the pool (via `close()`).

__Needs a migration guide entry__

artembilan commented:
This PR requires rebase and fixes for conflicts.
Thanks

artembilan commented on spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpInboundGateway.java:
```diff
@@ -70,29 +72,44 @@
+					this.serverConnectionFactory.closeConnection(connectionId);
+				}
+				else {
+					this.clientConnectionFactory.closeConnection(connectionId);
```
It doesn't relate to this PR, but looks like we really should do refactoring polishing around these `server/client` variables. I don't see so much different in usage here. 
So, the final code may be more simpler. 
Like background task :smile: 
Seems for me we haven't reached any scheduled background tasks yet...

garyrussell replied:
Yeah - we can probably have a single factory and a boolean to decide the type.

artembilan commented:
OK. Nothing to complain about!
I'm trying to fix conflicts my self and merging...

garyrussell commented:
No; I found a problem while fixing the conflicts.

artembilan commented:
OK. Thanks. Waiting and reviewing anything else :smile:

garyrussell committed:
INT-3654: Fix Late Listener Registration

Timing failures in `TcpOutboundGatewayTests.testFailoverCached()` (null listener).

The `FailoverClientConnectionFactory` registers its listener with the actual
connections it retrieves from a delegate.

When nesting failover and cached connection factories, we need to propagate the
`enableManualListenerRegistration` to the delegate factories so the connection
will wait for its listener to be registered.

garyrussell commented:
Rebased, pushed with fix for timing problem with nested Failover and Caching CFs.

artembilan commented:
Well, since I can't test it locally, I'll merge it as is, because review doesn't show any vulnerabilities.

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/ec5230a
